### ISSUE TITLE
[Discussion] Why are we letting Travis fail?

### ISSUE THREAD

**Issue by infinitewarp** [j-e-k/poketrainer#513]
Apologies in advance for soapboxing… 😰 

@j-e-k I just opened a pull request to fix a minor bug a short while ago, and when I discovered that the build was broken on my change, I realized we have been merging or committing straight to develop several times today while the build is failing. If people are committing *directly* to develop without opening pull requests from forks or other branches, well, it doesn't surprise me that the build would fail because it wouldn't have had an opportunity to run.

It looks like the develop branch broke by @stolencatkarma at https://github.com/j-e-k/poketrainer/commit/50621c1685d315d98e2fa600555311e8e05453cf — you can see the red **×**s in [the commit log](https://github.com/j-e-k/poketrainer/commits/develop). We've had several commits since it started failing, and nobody else stopped to fix the build.

Every contributor should be careful before hitting that "merge" button and *especially* before pushing commits directly to the master or develop branches. We're all responsible for keeping the code clean and the project healthy!

The build was easy enough to fix this time, and I just merged a commit that did, but we really shouldn't be ignoring Travis. He's our canary in the coal mine to stave off potential bugs. If we do ignore Travis when he yells at us, we should just disable the integration. (I've worked professionally in teams that had perpetually-broken CI, and IME, ignoring a broken CI can be as bad as or *worse* than having no CI at all.)

@stolencatkarma It looks like you committed and pushed directly to the develop branch with that commit. Is that right? If so, and if this comment https://github.com/j-e-k/poketrainer/commit/50621c1685d315d98e2fa600555311e8e05453cf suggests that was a mistake, since you have permission to push directly to the develop branch, you could have rebased and dropped the commit (or `git reset --hard HEAD~1` if it was your most recent one), and then force-pushed the branch back up to github to obliterate the commit. If you're quick about it, nobody would have seen the bogus change, and the repo wouldn't have unnecessary commits in its history. Technically, we could still do that now, but it would mess up the recent history, and anyone who has pulled/fetched since then would need to trash and clone again (or do a funky rebase dance).

/me steps off soapbox 🎉

**Comment by j-e-k** [comment]
This soapbox very much needed....thank you for being a voice of reason in this project

**Comment by j-e-k** [comment]
I know I just broke travis with the recent merge to master, I figured I would just get that merged given the new ban wave and then we can deal with code quality in the aftermath once we know more

**Comment by stolencatkarma** [comment]
Yeah. i'm very sorry about all that. I'm learning git as i go. there's a bug with the submit page for git that if you click "use a pull request" before the page finishes loading if will commit directly to the branch. 

If you ever need to get ahold of me directly to straighten me out on proper procedures i'm on the slack channel.

**Comment by infinitewarp** [comment]
@stolencatkarma Please do take some time to learn how to use forks, branches, rebases, and pull requests before making many more commits. I can understand eagerness to contribute ASAP, but following healthy development practices is essential to keeping open source projects stable (and the other contributors sane). Making changes directly to the files _on github.com itself_ especially in the _main_ repo is a really terrible idea and dangerous to the stability of the project. Except for the tiniest most absolutely urgent changes, nobody should make direct changes using the website's editing interface; I don't even do that for private personal projects.

(Unfortunately, my schedule is pretty spotty, and I don't have time to get on slack and wait around to chat synchronously. So, I'm writing up these notes to hopefully get us on track! Also I'm a huge proponent of transparency and just putting all this in the public records since others may learn from this exchange too. 😀 )

As a general policy, we should **always** be landing our commits via pull requests. Why?
1. Pull requests give visibility to other project contributors about what's coming down the pipe. Contributors can review your code and provide feedback and catch bugs before your code is running on other users' machines.
2. Pull requests automatically run the continuous integration jobs both when you open the request and with every subsequent push you make to the branch in that pull request. For this project, I set up a series of tasks for Travis CI to run to verify (roughly) the sanity of the code. This is what I'm referring to when I say Travis is angry. When the tasks in Travis fail, you see a message in the pull request on github with a red ×. _We should_ **_never**_ _merge a pull request when the Travis build is broken because this means the code may be broken!_
3. Pull requests allow you to change more than one file at a time.
4. Pull requests allow you to make multiple subsequent commits and then squash them together when it's time to merge and release to the world.
5. More info: https://help.github.com/categories/collaborating-on-projects-using-issues-and-pull-requests/

This is what failing CI integration looks like in a pull request:
![Travis mad. Travis smash.](http://i.imgur.com/A4PZ2HR.png)

To get yourself a good initial setup, you should:
1. create a fork of the main repo on github.com. https://help.github.com/articles/fork-a-repo/
2. clone your forked repo to your local computer. https://help.github.com/articles/cloning-a-repository/
3. create a remote upstream that points to the main repo. https://help.github.com/articles/configuring-a-remote-for-a-fork/

When you want to update your local copy after _other_ changes have landed in the main repo:
1. fetch the changes made upstream. https://help.github.com/articles/syncing-a-fork/
2. rebase your develop (and/or master) branch over upstream's branch. https://git-scm.com/book/en/v2/Git-Branching-Rebasing

When you want to make a change:
1. create a new branch from develop. https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging
2. commit your changes. https://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository
3. push your changes to _your_ remote fork on github. https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes#_pushing_remotes
4. open a pull request comparing your fork's branch to the main repo's develop branch. https://help.github.com/articles/creating-a-pull-request/
5. let the pull request run the CI jobs, give others a chance to see your change, have conversations about your changes in the thread, etc.

This probably sounds like a crapton of work at first, but it's really not. Once you get set up and understand the fundamentals, knocking through changes goes by quickly and becomes second nature. When I have a new request I want to open, it takes me only a few seconds to do all the necessary operations to get from my local computer into a pull request in the main repo.

Hope this helps! 🎉
